cpp/src/strings/convert/convert_urls.cu

```diff
@@ -185,137 +192,184 @@ constexpr bool is_hex_digit(char ch)
+  int local_warp_id     = threadIdx.x / cudf::detail::warp_size;
+  int warp_lane         = threadIdx.x % cudf::detail::warp_size;
+  int nwarps            = gridDim.x * blockDim.x / cudf::detail::warp_size;
+  char* in_chars_shared = temporary_buffer[local_warp_id];
+
+  // Loop through strings, and assign each string to a warp
+  for (size_type row_idx = global_warp_id; row_idx < num_strings; row_idx += nwarps) {
+    auto in_chars_string  = in_chars + in_offsets[row_idx];
+    auto out_chars_string = out_chars + out_offsets[row_idx];
+    auto string_length    = in_offsets[row_idx + 1] - in_offsets[row_idx];
+    int nblocks           = (string_length + block_size - 1) / block_size;
+
+    if (warp_lane == cudf::detail::warp_size - 1) { out_idx[local_warp_id] = 0; }
+
+    for (int block_idx = 0; block_idx < nblocks; block_idx++) {
+      int string_length_block = std::min(block_size, string_length - block_size * block_idx);
+
+      // Each warp collectively loads input characters of the current block to shared memory.
+      // Two halo cells with 0s before and after the input characters are added. The halo cells are
+      // used to test whether the current location as well as the previous two locations are escape
+      // characters, without branches.
+      for (int char_idx = warp_lane; char_idx < string_length_block + 4;
+           char_idx += cudf::detail::warp_size) {
+        char ch    = 0;
+        int in_idx = block_idx * block_size + char_idx - 2;
+        if (in_idx >= 0 && in_idx < string_length) ch = in_chars_string[in_idx];
+        in_chars_shared[char_idx] = ch;
       }
-    }
 
-    d_out_chars[output_idx] = ch;
-  }
-};
+      __syncwarp();
+
+      // `char_idx_start` represents the start character index of the current warp.
+      for (int char_idx_start = 0; char_idx_start < string_length_block;
+           char_idx_start += cudf::detail::warp_size) {
+        int char_idx = char_idx_start + warp_lane;
+        // The current thread should output a character if it is not part of an escape sequence
+        // starting at the previous two locations.
+        int8_t out_size =
+          (char_idx >= string_length_block || is_escape_char(in_chars_shared + char_idx) ||
+           is_escape_char(in_chars_shared + char_idx + 1))
+            ? 0
+            : 1;
+
+        // Warp-wise prefix sum to establish output location of the current thread.
+        // All threads in the warp participate in the prefix sum, even if `char_idx` is beyond
+        // `string_length_block`.
+        int8_t out_offset;
+        cub::WarpScan<int8_t>(cub_storage[local_warp_id]).ExclusiveSum(out_size, out_offset);
+
+        if (out_size == 1) {
+          char ch;
+          if (is_escape_char(in_chars_shared + char_idx + 2)) {
+            // If the current location is the start of an escape sequence, load and decode.
+            ch = (hex_char_to_byte(in_chars_shared[char_idx + 3]) << 4) |
```

Comment:
Is a bit shift by 4 standard for parsing escape sequences? I'm not familiar with how escape characters are parsed, so if this sequence of bitshift->bitwise or is standard it would be useful to put in another (`__device__ __forceinline__`) function so that it's behavior is very clear to non-string experts like me :)

Reply:
Addressed in https://github.com/rapidsai/cudf/pull/8622/commits/1e30938b5979055b582aaface178b0c357bd2018.